includes/derivatives.inc

```diff
@@ -0,0 +1,131 @@
+<?php
+
+/**
+ * @file
+ * Holds derivative functionality for compound solution pack.
+ */
+
+/**
+ * Derivative creation function for compound parent object TN.
+ *
+ * Wraps the TN update/creation function and
+ * adds expected logging.
+ *
+ * @param AbstractObject $object
+ *   The compound object.
+ * @param bool $force
+ *   Boolean for forcing this derivative creation.
+ *
+ * @return array
+ *   Success message.
+ */
+function islandora_compound_object_create_thumbnail(AbstractObject $object, $force = TRUE) {
+  if ($force || !isset($object['TN'])) {
+    $parts = islandora_compound_object_get_parts($object->id);
+    if (!empty($parts)) {
+      $child = islandora_object_load($parts[0]);
+      islandora_compound_object_update_parent_thumbnail($object, $child);
+    }
+    if (isset($child) && $object['TN']->size == $child['TN']->size) {
```

Comment:
I'm not sure this is being done at the appropriate place... If it was _before_ setting the content, it could be useful to avoid needlessly creating additional datastream versions, but _after_ just the check that the function did what it is designed... Could we not assume that `islandora_compound_object_update_parent_thumbnail()` would fail in some way, if the content is not set? As in, blowing up with an exception, or something of the like?

Really, that content-checking-bit beforehand could almost be shoved into the `islandora_compound_object_update_parent_thumbnail()`... would probably want to add something to the comment to indicate that it won't do anything if the datastream content is the same.